### ISSUE TITLE
Automated OTP Section Headings - Letters with Accents and html Entities Excluded from Anchor Link

### ISSUE THREAD

**Issue by lburack** [NCIOCPL/cgov-digital-platform#2343]
On Articles using automated On This Page (OTP) body sections, if the header includes a letter(s) with an accent mark(s), that letter is excluded from the anchor link text. This is an issue for Spanish.

For example, on https://www.cancer.gov/espanol/publicaciones/educacion-para-pacientes, the section "Exámenes de detección" creates an anchor link of "exmenes-de-deteccin" (https://www.cancer.gov/espanol/publicaciones/educacion-para-pacientes#exmenes-de-deteccin)

We should replace the characters with accent marks with their english character equivalent (e.g. the new anchor link would read "examenes-de-deteccion"

UPDATE 10/20/2023: This also applies to html entities. The content is `<h2>Some title&nbsp;</h2>` then the anchor becomes `some-title-nbsp`.

### Requirements

- All characters that are in other languages (e.g., Spanish letters, Greek letters, etc.) will be transliterated to their English equivalent
- HTML Entities will be removed from the url. See link [here](https://developer.mozilla.org/en-US/docs/Glossary/Entity) for applicable html entities
- All anchor text will be in lower case

### Acceptance Criteria

```gherkin
Scenario: A user views a url that has been generated from a title/heading that includes a letter in a language other than English
  Given the generated url was taken from a header/title that has a letter from a language other than English
  Then the letter will be transliterated to the English equivalent
    And the English transliteration will be put into the url
    And the user will see the url with the transliterated English equivalent
    And the transliterated English equivalent letter will be in lowercase
```

```gherkin
Scenario: A user views a url that has been generated from a title/heading that includes html entities
  Given the generated url was taken from a heading/title that has an html entity
  Then the html entity will be removed from the url
```

### Examples for testing
Accented letters in the Spanish alphabet with their equivalent replacement in a url are:

Á, á´= a; É, é = e; Í, í = i; Ó, ó = o; Ú, ú = u; Ñ, ñ = n; Ü, ü = u

Greek:

β = b

HTML Entity Example: "Chemotherapy & You" should have an anchor link of "chemotherapy-you"

**Comment by suryatheja-muppalla** [comment]
Request for more details.
Unable to update and save any existing node to check the behavior in localhost.
W.r.t URL: **http://www.devbox/espanol/node/138/edit**

Only after making sure the node is rightly getting updated, an action plan can be decided.

cc: @VictoriaSunNIH  @bryanpizzillo

**Comment by suryatheja-muppalla** [comment]
Tried to reproduce the issue, but the functionality is working fine in my localhost.

Below is the anchor link and query parameters that I have tested:
**<a href="#estudios-detección">Estudios detección</a>**
**Query parameter: #estudios-detección**

**<a href="#exmenes-de-detección">Exámenes de detección</a>**
**Query parameter: #exmenes-de-detección**

Needed to check:
Content editors should make sure the IDs are rightly entered inside anchor link.

**Comment by VictoriaSunNIH** [comment]
@suryatheja-muppalla - this isn't happening with IDs, as content editors don't create the on this page.  On this page is automated, for every body section that is added with a heading, any spanish characters in the heading that have an accent are stripped out of the on this page links that appear on the front end.  

See the links in the "En Esta Pagina" section of this page: https://www-cms-dev.cancer.gov/espanol/tipos/suprarrenal/pro/adultos/test

Both URLs are missing the characters that have accents.

**Comment by bryanpizzillo** [comment]
@suryatheja-muppalla please add a link to the code in GitHub that produces this anchor. Thanks!

**Comment by suryatheja-muppalla** [comment]
clean_id method is implemented in paragraph--body-section.html.twig and field--field-article-body.html.twig files.
Actually ‘clean_id’ method is responsible to generate string with Drupal machine name standards (space replaced with Underscore and other character sanitization’s are performed).In the process of sanitization , it is removing few Spanish characters, that is creating the issue.

I am trying to perform string replace of Spanish characters with relevant English characters in .theme file using ‘cgov_common_preprocess_field__field_article_body’ method.

I am trying to access the updated string in paragraph--body-section.html.twig and field--field-article-body.html.twig files currently and generate the anchor links.

**Comment by navsunka** [comment]
@bryanpizzillo , shall we add a new twig filter for parsing accents with below logic
iconv('UTF-8', 'US-ASCII//TRANSLIT', $string), 

Case1:
<?php
echo iconv('UTF-8', 'US-ASCII//TRANSLIT', 'A æ Übérmensch på høyeste nivå! И я люблю PHP! есть. ﬁ');
?>
**A ae Ubermensch pa hoyeste niva! ? ? ????? PHP! ????. fi**

Case2:
If we enable new new php extension php-intl in the server we will get output like below
var_dump(iconv("UTF-8", "ASCII//TRANSLIT//IGNORE", transliterator_transliterate('Any-Latin; Latin-ASCII',
    "A æ Übérmensch på høyeste nivå! И я люблю PHP! есть. ﬁ"));
**A ae Ubermensch pa hoyeste niva! I a lublu PHP! est'. fi**

**Comment by laurelthrash** [comment]
Naveen to pick up work on this ticket following completion of active work. Andy to work on test cases for this ticket.

**Comment by laurelthrash** [comment]
Note from SU: @navsunka running tests locally. Will add to ODE and create a new PR once complete.
